source/base/ThreadPool.ooc

```diff
@@ -1,111 +1,203 @@
 use ooc-collections
+use ooc-base
 
 import structs/LinkedList
 import threading/Thread
+import threading/native/ConditionUnix
+import os/Time
 
-ThreadJob: class {
-	_body: Func
-	init: func (=_body)
-	_execute: func { this _body() }
-	run: virtual func {
-		this _execute()
-		this free()
-	}
+_TaskState: enum {
+	Unfinished
+	Finished
+	Cancelled
 }
 
-SynchronizedThreadJob: class extends ThreadJob {
-	_finishedCondition := WaitCondition new()
-	_mutex := Mutex new()
-	_finished := false
-	finished ::= this _finished
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
+_Task: class {
+	_mutex: Mutex
+	_action: Func
```

Comment:
This should not be in this class. My suggestion is to create an abstract class without any function pointers, and then you inherit two classes from it: _ActionTask and _ResultTask where the different function pointers are located.